qutebrowser/browser/tab.py

```diff
@@ -0,0 +1,535 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base class for a wrapper over QWebView/QWebEngineView."""
+
+import itertools
+
+from PyQt5.QtCore import pyqtSignal, pyqtSlot, QUrl, QObject, QPoint
+from PyQt5.QtGui import QIcon
+from PyQt5.QtWidgets import QWidget, QLayout
+
+from qutebrowser.config import config
+from qutebrowser.utils import utils, objreg, usertypes, message
+
+
+tab_id_gen = itertools.count(0)
+
+
+Backend = usertypes.enum('Backend', ['QtWebKit', 'QtWebEngine'])
+
+
+class WebTabError(Exception):
+
+    """Base class for various errors."""
+
+
+class WrapperLayout(QLayout):
+
+    """A Qt layout which simply wraps a single widget.
+
+    This is used so the widget is hidden behind a AbstractTab API and can't
+    easily be accidentally accessed.
+    """
+
+    def __init__(self, widget, parent=None):
+        super().__init__(parent)
+        self._widget = widget
+
+    def addItem(self, _widget):
+        raise AssertionError("Should never be called!")
+
+    def sizeHint(self):
+        return self._widget.sizeHint()
+
+    def itemAt(self, _index):  # pragma: no cover
+        # For some reason this sometimes gets called by Qt.
+        return None
+
+    def takeAt(self, _index):
+        raise AssertionError("Should never be called!")
+
+    def setGeometry(self, rect):
+        self._widget.setGeometry(rect)
+
+
+class TabData:
+
+    """A simple namespace with a fixed set of attributes.
+
+    Attributes:
+        _initializing: Set when we're currently in __init__.
+        keep_icon: Whether the (e.g. cloned) icon should not be cleared on page
+                   load.
+        inspector: The QWebInspector used for this webview.
+        viewing_source: Set if we're currently showing a source view.
+    """
+
+    def __init__(self):
+        self._initializing = True
+        self.keep_icon = False
+        self.viewing_source = False
+        self.inspector = None
+        self._initializing = False
+
+    def __setattr__(self, attr, value):
```

Comment:
Nice - with `__slots__`, pylint even statically checks the attributes being set and complains :+1: